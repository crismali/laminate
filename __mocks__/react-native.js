/* eslint-env jest */

class Value {
  constructor (value) {
    this.value = value
  }

  interpolate (args) {
    return { interpolated: this.value, args: args }
  }
}

const AnimatedView = ({ children }) => children
AnimatedView.displayName = 'Animated.View'

const animatedTimingStart = jest.fn()
const Animated = {
  Value,
  View: AnimatedView,
  timing: jest.fn(() => ({ start: animatedTimingStart })),
  animatedTimingStart
}

const Dimensions = { get: jest.fn(() => ({ width: 375, height: 667 })) }

const PanResponder = { create: jest.fn((args) => ({ panResponder: args })) }

const View = ({ children }) => children

const StyleSheet = { create: jest.fn(styles => styles) }

export {
  Animated,
  Dimensions,
  PanResponder,
  StyleSheet,
  View
}
