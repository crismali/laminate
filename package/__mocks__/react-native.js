/* eslint-env jest */
import React from 'react'

class Value {
  constructor (value) {
    this.value = value
  }

  interpolate (args) {
    return { interpolated: this.value, args: args }
  }

  setValue (value) {
    this.value = value
  }
}

const AnimatedView = ({ children }) => children
AnimatedView.displayName = 'Animated.View'

const animatedTimingStart = jest.fn((func) => func && func())
const Animated = {
  Value,
  View: AnimatedView,
  timing: jest.fn(() => ({ start: animatedTimingStart })),
  animatedTimingStart
}

const Dimensions = { get: jest.fn(() => ({ width: 375, height: 667 })) }

const Easing = { inOut: jest.fn() }

const PanResponder = { create: jest.fn((args) => ({ panHandlers: args })) }

const View = ({ children }) => <div>{children}</div> // eslint-disable-line

const StyleSheet = { create: jest.fn(styles => styles) }

export {
  Animated,
  Dimensions,
  Easing,
  PanResponder,
  StyleSheet,
  View
}
