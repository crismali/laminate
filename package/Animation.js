import React, { Component } from 'react'
import { Animated, Easing } from 'react-native'
import PropTypes from 'prop-types'

const { Provider, Consumer } = React.createContext()

export class Animation extends Component {
  static defaultProps = {
    duration: 375,
    easing: Easing.inOut(Easing.cubic)
  }

  static propTypes = {
    animation: PropTypes.instanceOf(Animated.Value),
    children: PropTypes.any.isRequired,
    duration: PropTypes.number,
    easing: PropTypes.any
  }

  state = {
    animation: new Animated.Value(0)
  }

  animationProps = () => {
    return {
      play: this.play,
      rewind: this.rewind,
      value: this.state.animation
    }
  }

  play = (options = {}) => {
    this.animate({ toValue: 1, ...options })
  }

  rewind = (options = {}) => {
    this.animate({ toValue: 0, ...options })
  }

  animate = ({ toValue, duration, easing, callback }) => {
    Animated.timing(this.state.animation, {
      toValue,
      duration: duration || this.props.duration,
      easing: easing || this.props.easing
    }).start(() => {
      callback && callback()
    })
  }

  children = () => {
    const { children } = this.props

    if (typeof children === 'function') {
      return <Consumer>{children}</Consumer>
    } else {
      return children
    }
  }

  render () {
    return <Provider value={this.animationProps()}>
      {this.children()}
    </Provider>
  }
}

export const WithAnimation = Consumer
