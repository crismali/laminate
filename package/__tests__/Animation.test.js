/* eslint-env jest */
import React from 'react'
import { Animated } from 'react-native'
import { shallow } from 'enzyme'
import { Animation } from '../Animation'

const Child = () => 'a child'

describe('state', () => {
  it('creates a new animated value for the animation state', () => {
    const subject = shallow(<Animation><Child /></Animation>)
    const result = subject.state('animation')
    expect(result).toBeInstanceOf(Animated.Value)
    expect(result.value).toEqual(0)
  })
})

describe('#animationProps', () => {
  it('has animation with play, rewind, and value', () => {
    const subject = shallow(<Animation><Child /></Animation>)

    expect(subject.instance().animationProps()).toEqual({
      play: subject.instance().play,
      rewind: subject.instance().rewind,
      value: subject.state('animation')
    })
  })
})

describe('#play', () => {
  it('animates to one by default', () => {
    const subject = shallow(<Animation><Child /></Animation>)
    subject.instance().animate = jest.fn()
    subject.instance().play()
    expect(subject.instance().animate).toHaveBeenCalledWith({ toValue: 1 })
  })

  it('passes through arguments', () => {
    const subject = shallow(<Animation><Child /></Animation>)
    subject.instance().animate = jest.fn()
    subject.instance().play({ toValue: 5, foo: 'bar' })
    expect(subject.instance().animate).toHaveBeenCalledWith({ toValue: 5, foo: 'bar' })
  })
})

describe('#rewind', () => {
  it('animates to zero by default', () => {
    const subject = shallow(<Animation><Child /></Animation>)
    subject.instance().animate = jest.fn()
    subject.instance().rewind()
    expect(subject.instance().animate).toHaveBeenCalledWith({ toValue: 0 })
  })

  it('passes through arguments', () => {
    const subject = shallow(<Animation><Child /></Animation>)
    subject.instance().animate = jest.fn()
    subject.instance().play({ toValue: 5, foo: 'bar' })
    expect(subject.instance().animate).toHaveBeenCalledWith({ toValue: 5, foo: 'bar' })
  })
})

describe('#animate', () => {
  it('runs the animation', () => {
    const subject = shallow(<Animation><Child /></Animation>)
    subject.instance().animate({ toValue: 5, duration: 20, easing: 'something' })
    expect(Animated.timing).toHaveBeenCalledWith(subject.state('animation'), {
      toValue: 5,
      duration: 20,
      easing: 'something'
    })
  })

  it('runs the callback', () => {
    const subject = shallow(<Animation><Child /></Animation>)
    const callback = jest.fn()
    subject.instance().animate({ toValue: 5, callback })
    expect(callback).toHaveBeenCalled()
  })

  it('falls back to the easing and duration props when not provided', () => {
    const subject = shallow(<Animation duration={14} easing='so easy'><Child /></Animation>)
    subject.instance().animate({ toValue: 53 })
    expect(Animated.timing).toHaveBeenCalledWith(subject.state('animation'), {
      toValue: 53,
      duration: 14,
      easing: 'so easy'
    })
  })
})

describe('#render', () => {
  it('renders a children function with arguments', () => {
    const subject = shallow(<Animation>
      {animation => (
        <Child animation={animation} />
      )}
    </Animation>)
    expect(subject).toMatchSnapshot()
  })

  it('renders a regular children', () => {
    const subject = shallow(<Animation><Child /></Animation>)
    expect(subject).toMatchSnapshot()
  })
})
