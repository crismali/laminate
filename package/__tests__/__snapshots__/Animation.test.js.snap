// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#render renders a children function with arguments 1`] = `
<[object Object]
  value={
    Object {
      "play": [Function],
      "rewind": [Function],
      "value": Value {
        "value": 0,
      },
    }
  }
>
  <[object Object] />
</[object Object]>
`;

exports[`#render renders a regular children 1`] = `
<[object Object]
  value={
    Object {
      "play": [Function],
      "rewind": [Function],
      "value": Value {
        "value": 0,
      },
    }
  }
>
  <Child />
</[object Object]>
`;
